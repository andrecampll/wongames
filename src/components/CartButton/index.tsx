import {
  AddShoppingCart,
  RemoveShoppingCart,
} from '@styled-icons/material-outlined';

import Button, { ButtonProps } from '../Button';
import { useCart } from '../../hooks/cart/useCart';

type CartButtonProps = {
  id: string;
  hasText?: boolean;
} & Pick<ButtonProps, 'size'>;

export const CartButton = ({ id, size, hasText }: CartButtonProps) => {
  const { isInCart, addToCart, removeFromCart } = useCart();
  const ButtonText = isInCart(id) ? 'Remove from cart' : 'Add to cart';

  return (
    <Button
      icon={
        isInCart(id) ? (
          <RemoveShoppingCart aria-label="Remove from cart" />
        ) : (
          <AddShoppingCart aria-label="Add to cart" />
        )
      }
      size={size}
      onClick={() => (isInCart(id) ? removeFromCart(id) : addToCart(id))}
    >
      {hasText && ButtonText}
    </Button>
  );
};
