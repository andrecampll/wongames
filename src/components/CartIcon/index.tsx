import { ShoppingCart } from '@styled-icons/material-outlined/ShoppingCart';
import { useCart } from '../../hooks/cart/useCart';

import { Wrapper, Badge } from './styles';

const CartIcon = () => {
  const { quantity } = useCart();

  return (
    <Wrapper>
      {quantity > 0 && <Badge aria-label="Cart Items">{quantity}</Badge>}
      <ShoppingCart aria-label="Shopping Cart" />
    </Wrapper>
  );
};

export default CartIcon;
