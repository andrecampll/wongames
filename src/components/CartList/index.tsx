import Link from 'next/link';
import { useCart } from '../../hooks/cart/useCart';

import Button from '../Button';
import Empty from '../Empty';
import GameItem from '../GameItem';
import Loader from '../Loader';

import { Wrapper, Footer, Total, Loading, GamesList } from './styles';

export type CartListProps = {
  hasButton?: boolean;
};

const CartList = ({ hasButton = false }: CartListProps) => {
  const { items, total, loading } = useCart();

  if (loading) {
    return (
      <Loading>
        <Loader />
      </Loading>
    );
  }

  return (
    <Wrapper isEmpty={!items.length}>
      {items.length ? (
        <>
          <GamesList>
            {items.map(item => (
              <GameItem key={item.title} {...item} />
            ))}
          </GamesList>

          <Footer>
            {!hasButton && <span>Total:</span>}
            <Total>{total}</Total>

            {hasButton && (
              <Link href="/cart">
                <Button as="a">Buy it now</Button>
              </Link>
            )}
          </Footer>
        </>
      ) : (
        <Empty
          title="Your cart is empty"
          description="Go back to the store and explore great games and offers."
          hasLink
        />
      )}
    </Wrapper>
  );
};

export default CartList;
