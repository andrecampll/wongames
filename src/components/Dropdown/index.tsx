import { useState } from 'react';

import { Wrapper, Title, Content, Overlay } from './styles';

export type DropdownProps = {
  title: React.ReactNode;
  children: React.ReactNode;
};

const Dropdown = ({ title, children }: DropdownProps) => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <Wrapper isOpen={isOpen}>
      <Title onClick={() => setIsOpen(!isOpen)}>{title}</Title>

      <Content aria-hidden={!isOpen}>{children}</Content>
      <Overlay aria-hidden={!isOpen} onClick={() => setIsOpen(!isOpen)} />
    </Wrapper>
  );
};

export default Dropdown;
