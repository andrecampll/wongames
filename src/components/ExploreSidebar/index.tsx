import { useEffect, useState } from 'react';
import xor from 'lodash.xor';
import { Close } from '@styled-icons/material-outlined/Close';
import { FilterList } from '@styled-icons/material-outlined/FilterList';

import { ParsedUrlQueryInput } from 'querystring';
import Heading from '../Heading';
import Checkbox from '../Checkbox';
import Radio from '../Radio';
import Button from '../Button';

import {
  Wrapper,
  Overlay,
  IconWrapper,
  Items,
  Content,
  Footer,
} from './styles';

export type ItemProps = {
  title: string;
  name: string;
  type: string;
  fields: Field[];
};

type Field = {
  label: string;
  name: string;
};

type Values = ParsedUrlQueryInput;

export type ExploreSidebarProps = {
  items: ItemProps[];
  initialValues?: Values;
  onFilter: (values: Values) => void;
};

const ExploreSidebar = ({
  items,
  initialValues = {},
  onFilter,
}: ExploreSidebarProps) => {
  const [values, setValues] = useState(initialValues);
  const [isOpen, setIsOpen] = useState(false);

  useEffect(() => {
    onFilter(values);
    // Thist method comes from another template
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [values]);

  const handleCheckbox = (name: string, value: string) => {
    const currentList = (values[name] as []) || [];
    setValues(s => ({ ...s, [name]: xor(currentList, [value]) }));
  };

  const handleRadio = (name: string, value: string | boolean) => {
    setValues(s => ({ ...s, [name]: value }));
  };

  const handleFilterMenu = () => {
    setIsOpen(false);
  };

  return (
    <Wrapper isOpen={isOpen}>
      <Overlay aria-hidden={isOpen} />
      <IconWrapper>
        <FilterList aria-label="open filters" onClick={() => setIsOpen(true)} />
        <Close aria-label="close filters" onClick={() => setIsOpen(false)} />
      </IconWrapper>

      <Content>
        {items.map(item => (
          <Items key={item.title}>
            <Heading lineBottom lineColor="secondary" size="small">
              {item.title}
            </Heading>

            {item.type === 'checkbox' &&
              item.fields.map(field => (
                <Checkbox
                  key={field.name}
                  name={field.name}
                  label={field.label}
                  labelFor={field.name}
                  isChecked={(values[item.name] as string[])?.includes(
                    field.name,
                  )}
                  onCheck={() => handleCheckbox(item.name, field.name)}
                />
              ))}

            {item.type === 'radio' &&
              item.fields.map(field => (
                <Radio
                  id={field.name}
                  key={field.name}
                  value={field.name}
                  name={item.name}
                  label={field.label}
                  labelFor={field.name}
                  defaultChecked={
                    String(field.name) === String(values[item.name])
                  }
                  onChange={() => handleRadio(item.name, field.name)}
                />
              ))}
          </Items>
        ))}
      </Content>

      <Footer>
        <Button fullWidth size="medium" onClick={handleFilterMenu}>
          Filter
        </Button>
      </Footer>
    </Wrapper>
  );
};

export default ExploreSidebar;
