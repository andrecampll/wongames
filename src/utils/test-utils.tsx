import { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';

import { ThemeProvider } from 'styled-components';

import {
  WishlistContext,
  WishlistContextDefaultValues,
  WishlistContextDTO,
} from '../hooks/wishlist/useWishlist';

import theme from '../styles/theme';
import {
  CartContext,
  CartContextDTO,
  CartContextDefaultValues,
} from '../hooks/cart/useCart';

type CustomRenderProps = {
  cartProviderProps?: CartContextDTO;
  wishlistProviderProps?: WishlistContextDTO;
} & Omit<RenderOptions, 'queries'>;

const customRender = (
  ui: ReactElement,
  {
    cartProviderProps = CartContextDefaultValues,
    wishlistProviderProps = WishlistContextDefaultValues,
    ...renderOptions
  }: CustomRenderProps = {},
) =>
  render(
    <ThemeProvider theme={theme}>
      <CartContext.Provider value={cartProviderProps}>
        <WishlistContext.Provider value={wishlistProviderProps}>
          {ui}
        </WishlistContext.Provider>
      </CartContext.Provider>
    </ThemeProvider>,
    renderOptions,
  );

export * from '@testing-library/react';
export { customRender as render };
