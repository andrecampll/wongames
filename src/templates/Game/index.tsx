/* eslint-disable jsx-a11y/aria-role */
import Base from '../Base';

import GameInfo, { GameInfoProps } from '../../components/GameInfo';
import Gallery, { GalleryImageProps } from '../../components/Gallery';
import TextContent from '../../components/TextContent';
import GameDetails, { GameDetailsProps } from '../../components/GameDetails';
import { GameCardProps } from '../../components/GameCard';
import { HighlightProps } from '../../components/Highlight';
import Showcase from '../../components/Showcase';
import { Divider } from '../../components/Divider';

import {
  Cover,
  Main,
  SectionGameInfo,
  SectionGallery,
  SectionDescription,
  SectionGameDetails,
} from './styles';

export type GameTemplateProps = {
  cover: string;
  gameInfo: GameInfoProps;
  gallery?: GalleryImageProps[];
  description: string;
  details: GameDetailsProps;
  upcomingTitle: string;
  upcomingGames: GameCardProps[];
  upcomingHighlight: HighlightProps;
  recommendedGames: GameCardProps[];
};

const Game = ({
  cover,
  gameInfo,
  gallery,
  description,
  details,
  upcomingTitle,
  upcomingGames,
  upcomingHighlight,
  recommendedGames,
}: GameTemplateProps) => (
  <Base>
    <Cover src={cover} role="image" aria-label="cover" />

    <Main>
      <SectionGameInfo>
        <GameInfo {...gameInfo} />
      </SectionGameInfo>

      <SectionGallery>
        {!!gallery && <Gallery items={gallery} />}
      </SectionGallery>

      <SectionDescription>
        <TextContent title="Description" content={description} />
      </SectionDescription>

      <SectionGameDetails>
        <GameDetails {...details} />
        <Divider />
      </SectionGameDetails>

      <Showcase
        title={upcomingTitle}
        games={upcomingGames}
        highlight={upcomingHighlight}
      />

      <Showcase title="You may like these games" games={recommendedGames} />
    </Main>
  </Base>
);

export default Game;
