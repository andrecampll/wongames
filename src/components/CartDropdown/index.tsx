import CartIcon from '../CartIcon';
import Dropdown from '../Dropdown';
import CartList from '../CartList';

import { Wrapper } from './styles';

const CartDropdown = () => (
  <Wrapper>
    <Dropdown title={<CartIcon />}>
      <CartList hasButton />
    </Dropdown>
  </Wrapper>
);

export default CartDropdown;
