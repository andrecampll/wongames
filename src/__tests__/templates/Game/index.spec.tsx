import 'jest-styled-components';
import { screen } from '@testing-library/react';
import { render } from '../../../utils/test-utils';

import galleryMock from '../../../components/Gallery/mock';
import gameInfoMock from '../../../components/GameInfo/mock';
import gameDetailsMock from '../../../components/GameDetails/mock';
import gamesMock from '../../../components/GameCardSlider/mock';
import highlightMock from '../../../components/Highlight/mock';

import Game, { GameTemplateProps } from '../../../templates/Game';
import { GameDetailsProps } from '../../../components/GameDetails';

const props: GameTemplateProps = {
  cover: 'bg-image.jpg',
  gameInfo: gameInfoMock,
  gallery: galleryMock,
  description: `<h1>Custom HTML</h1>`,
  details: gameDetailsMock as GameDetailsProps,
  upcomingGames: gamesMock,
  upcomingHighlight: highlightMock,
  recommendedGames: gamesMock,
  upcomingTitle: 'Upcoming games',
};

jest.mock('templates/Base', () => ({
  __esModule: true,
  default: function Mock({ children }: { children: React.ReactNode }) {
    return <div data-testid="Mock Base">{children}</div>;
  },
}));

jest.mock('../../../components/Menu', () => {
  return {
    __esModule: true,
    default: function Mock() {
      return <div data-testid="Mock Menu" />;
    },
  };
});

jest.mock('../../../components/Gallery', () => ({
  __esModule: true,
  default: function Mock() {
    return <div data-testid="Mock Gallery" />;
  },
}));

jest.mock('../../../components/GameDetails', () => ({
  __esModule: true,
  default: function Mock() {
    return <div data-testid="Mock GameDetails" />;
  },
}));

jest.mock('../../../components/GameInfo', () => ({
  __esModule: true,
  default: function Mock() {
    return <div data-testid="Mock GameInfo" />;
  },
}));

jest.mock('../../../components/Showcase', () => ({
  __esModule: true,
  default: function Mock() {
    return <div data-testid="Mock Showcase" />;
  },
}));

describe('Game', () => {
  it('should be able to render with components', () => {
    const { container } = render(<Game {...props} />);

    expect(screen.getByTestId('Mock Gallery')).toBeInTheDocument();
    expect(screen.getByTestId('Mock GameDetails')).toBeInTheDocument();
    expect(screen.getByTestId('Mock GameInfo')).toBeInTheDocument();
    expect(screen.getAllByTestId('Mock Showcase')).toHaveLength(2);
    expect(screen.getByText(/custom html/i)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('shout not render the gallery if has no image', () => {
    render(<Game {...props} gallery={undefined} />);

    expect(screen.queryByTestId('Mock Gallery')).not.toBeInTheDocument();
  });

  it('shout not render the gallery on mobile', () => {
    render(<Game {...props} />);

    expect(screen.getByTestId('Mock Gallery').parentElement).toHaveStyle({
      display: 'none',
    });

    expect(screen.getByTestId('Mock Gallery').parentElement).toHaveStyleRule(
      'display',
      'block',
      {
        media: '(min-width: 768px)',
      },
    );
  });

  it('should render the cover image', () => {
    render(<Game {...props} />);

    const cover = screen.getByRole('image', { name: /cover/i });

    expect(cover).toHaveStyle({
      backgroundImage: 'url(bg-image.jpg)',
      height: '39.5rem',
    });

    expect(cover).toHaveStyleRule('height', '70rem', {
      media: '(min-width: 768px)',
    });

    expect(cover).toHaveStyleRule(
      'clip-path',
      'polygon(0 0,100% 0,100% 100%,0 85%)',
      {
        media: '(min-width: 768px)',
      },
    );
  });
});
