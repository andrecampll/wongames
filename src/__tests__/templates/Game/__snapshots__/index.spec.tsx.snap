// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should be able to render with components 1`] = `
.c7 {
  color: #FAFAFA;
  padding-left: 0.8rem;
  border-left: 0.7rem solid #3CD3C1;
  font-size: 2.0rem;
}

.c6 {
  color: #FAFAFA;
}

.c6 h1,
.c6 h2,
.c6 h3,
.c6 h4,
.c6 h5,
.c6 h6 {
  color: #FAFAFA;
  margin: 1.6rem 0;
  padding-left: 0.8rem;
  border-left: 0.7rem solid #3CD3C1;
}

.c6 p {
  margin-bottom: 1.6rem;
}

.c6 a {
  color: #F231A5;
}

.c6 img {
  max-width: min(70rem,100%);
  margin-bottom: 1.6rem;
}

.c6 ul,
.c6 ol {
  padding: 1.6rem 2.4rem;
}

.c6 hr {
  margin: 2.4rem 0;
}

.c8 {
  margin: 5.6rem auto 3.2rem;
  height: 0.1rem;
  background: rgba(181,181,181,0.3);
  border: 0;
}

.c2 {
  width: 100%;
  max-width: 130rem;
  margin-left: auto;
  margin-right: auto;
  padding-left: calc(3.2rem / 2);
  padding-right: calc(3.2rem / 2);
}

.c1 {
  margin-top: 20rem;
}

.c0 {
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  height: 39.5rem;
  background-image: url(bg-image.jpg);
  background-size: cover;
  background-position: top center;
  opacity: 0.4;
}

.c3 {
  margin-bottom: 4.8rem;
}

.c4 {
  display: none;
}

.c5 .description__copyrights {
  color: #8F8F8F;
  font-size: 1.2rem;
  margin-top: 3.2rem;
}

@media (min-width:768px) {
  .c7 {
    font-size: 2.8rem;
  }
}

@media (min-width:768px) {
  .c6 {
    background: #FAFAFA;
    padding: 3.2rem;
    color: #030517;
  }
}

@media (min-width:768px) {
  .c6 h1,
  .c6 h2,
  .c6 h3,
  .c6 h4,
  .c6 h5,
  .c6 h6 {
    color: #030517;
  }
}

@media (min-width:768px) {
  .c8 {
    margin: calc(5.6rem * 2.5) auto 5.6rem;
  }
}

@media (min-width:768px) {
  .c1 {
    margin-top: 45rem;
  }
}

@media (min-width:768px) {
  .c0 {
    height: 70rem;
    -webkit-clip-path: polygon(0 0,100% 0,100% 100%,0 85%);
    clip-path: polygon(0 0,100% 0,100% 100%,0 85%);
  }
}

@media (min-width:768px) {
  .c3 {
    margin-bottom: calc(4.8rem * 2);
  }
}

@media (min-width:768px) {
  .c4 {
    display: block;
  }
}

<div
  data-testid="Mock Base"
>
  <div
    aria-label="cover"
    class="c0"
    role="image"
    src="bg-image.jpg"
  />
  <main
    class="c1"
  >
    <section
      class="c2 c3"
    >
      <div
        data-testid="Mock GameInfo"
      />
    </section>
    <section
      class="c2 c3 c4"
    >
      <div
        data-testid="Mock Gallery"
      />
    </section>
    <section
      class="c2 c3 c5"
    >
      <div
        class="c6"
      >
        <h2
          class="c7"
          color="white"
        >
          Description
        </h2>
        <div>
          <h1>
            Custom HTML
          </h1>
        </div>
      </div>
    </section>
    <section
      class="c2 c3"
    >
      <div
        data-testid="Mock GameDetails"
      />
      <hr
        class="c8"
      />
    </section>
    <div
      data-testid="Mock Showcase"
    />
    <div
      data-testid="Mock Showcase"
    />
  </main>
</div>
`;
